### PR TITLE
[Universal] Single CommandBuffer for URP

### PR DESCRIPTION
**Pull request by jonuuukas**
This PR changes how we use CommandBuffers inside URP.

Previously, we used a lot of local cmd buffers across various places in URP to execute commands. This PR changes that by adding commandBuffer to RenderingData struct which is passed to every place, where command buffer is needed. This also eliminates the need to manually call ExecuteCommandBuffer and CommandBufferPool.Release() methods as they will be automatically called after each ScriptableRenderPass::Execute() call. 

This is not a disruptive change to the pipeline or the API as no API is getting added, it's only a minor disruption on creating new passes, as you cannot create your own cmd buffers anymore (not technically prohibited, as there is no way to catch that).

URP functionality-wise this changes nothing, this is basically a refactor, although touching a lot of our code

TDD can be found [here](https://docs.google.com/document/d/1q4C-wJgCPMcufYhNrOfcLdSIv9loeWhnZuOOWASIJ3I/edit?usp=sharing)

Manually ran OSX 2D and Foundation tests. Also checked UniversalRendererExamples on OSX Editor and iOS

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs [hunk 1]
```diff
@@ -89,7 +89,7 @@ private void CopyCameraSortingLayerRenderTexture(ScriptableRenderContext context
                 ClearFlag.None, Color.clear);
             cmd.SetGlobalTexture(k_CameraSortingLayerTextureID, m_Renderer2DData.cameraSortingLayerRenderTarget.nameID);
             context.ExecuteCommandBuffer(cmd);
-            CommandBufferPool.Release(cmd);
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:92]: can ExecuteCommandBuffer be removed from here since it happens also later? Or it's important that Execute and Clear happen here?

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs [hunk 2]
```diff
@@ -162,7 +162,8 @@ public override void Setup(ScriptableRenderContext context, ref RenderingData re
                 {
                     stackHasPostProcess = stackHasPostProcess && DebugHandler.IsPostProcessingAllowed;
                 }
-                DebugHandler.Setup(context, ref cameraData);
+
+                DebugHandler.Setup(context, ref cameraData, renderingData.commandBuffer);
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs:166]: should we pass RenderingData instead of CommandBuffer for consistency with other methods? Also passing RenderingData should be more future proof in case more info (already in renderingData) would be required in the future by the function

File: com.unity.render-pipelines.universal/Runtime/Debug/DebugHandler.cs [hunk 3]
```diff
@@ -281,7 +279,7 @@ internal void Setup(ScriptableRenderContext context, ref CameraData cameraData)
             }
 
             context.ExecuteCommandBuffer(cmd);
-            CommandBufferPool.Release(cmd);
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/Debug/DebugHandler.cs:284]: similar question as before: are ExecuteCommandBuffer and Clear needed because of cmd buffer syncing?

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs [hunk 4]
```diff
@@ -377,7 +377,7 @@ public void Setup(ScriptableRenderContext context, ref RenderingData renderingDa
                 m_LightCookieManager.Setup(context, cmd, ref renderingData.lightData);
             }
             context.ExecuteCommandBuffer(cmd);
-            CommandBufferPool.Release(cmd);
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/ForwardLights.cs:380]: since I see the pattern that all the setup methods need the explicit ExecuteCommandBuffer and Clear at the end. Could we call this once from ScriptableRenderer, so we just flush the commands in between Setup and Execute, and every pass implementation doesn't have to do it manually every time?

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs [hunk 5]
```diff
@@ -828,12 +828,12 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
         {
             if (m_CreateEmptyShadowmap)
             {
-                SetEmptyAdditionalShadowmapAtlas(ref context);
+                SetEmptyAdditionalShadowmapAtlas(ref context, renderingData.commandBuffer);
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:831]: should we pass RenderingData instead of cmd?

Comment (manuele-bonanno) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:831]: that would also be consistent with RenderAdditionalShadowmapAtlas

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs [hunk 6]
```diff
@@ -852,9 +852,8 @@ void Clear()
             m_GlobalShadowSliceIndexToPerLightShadowSliceIndex.Clear();
         }
 
-        void SetEmptyAdditionalShadowmapAtlas(ref ScriptableRenderContext context)
+        void SetEmptyAdditionalShadowmapAtlas(ref ScriptableRenderContext context, CommandBuffer cmd)
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs:855]: should we pass RenderingData instead of cmd?

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs [hunk 7]
```diff
@@ -173,21 +173,24 @@ void Clear()
                 m_CascadeSlices[i].Clear();
         }
 
-        void SetEmptyMainLightCascadeShadowmap(ref ScriptableRenderContext context)
+        void SetEmptyMainLightCascadeShadowmap(ref ScriptableRenderContext context, CommandBuffer cmd)
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:176]: should we pass RenderingData instead of cmd?

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs [hunk 8]
```diff
@@ -173,21 +173,24 @@ void Clear()
...
                 new Vector4(1, 0, 1, 0));
             cmd.SetGlobalVector(MainLightShadowConstantBuffer._ShadowmapSize,
                 new Vector4(1f / m_EmptyLightShadowmapTexture.rt.width, 1f / m_EmptyLightShadowmapTexture.rt.height, m_EmptyLightShadowmapTexture.rt.width, m_EmptyLightShadowmapTexture.rt.height));
             context.ExecuteCommandBuffer(cmd);
-            CommandBufferPool.Release(cmd);
+            cmd.Clear();
         }
 
-        void RenderMainLightCascadeShadowmap(ref ScriptableRenderContext context, ref CullingResults cullResults, ref LightData lightData, ref ShadowData shadowData)
+        void RenderMainLightCascadeShadowmap(ref ScriptableRenderContext context, /*ref CullingResults cullResults, ref LightData lightData, ref ShadowData shadowData,*/ ref RenderingData renderingData)
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs:188]: looks like you forgot the commented out code

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 9]
```diff
@@ -1588,13 +1590,12 @@ static void SetRenderTarget(CommandBuffer cmd, RTHandle[] colorAttachments, RTHa
         internal virtual void EnableSwapBufferMSAA(bool enable) { }
 
         [Conditional("UNITY_EDITOR")]
-        void DrawGizmos(ScriptableRenderContext context, Camera camera, GizmoSubset gizmoSubset)
+        void DrawGizmos(ScriptableRenderContext context, Camera camera, GizmoSubset gizmoSubset, CommandBuffer cmd)
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:1593]: should we pass RenderingData instead of cmd?

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs [hunk 10]
```diff
@@ -993,7 +992,7 @@ static void InitializeAdditionalCameraData(Camera camera, UniversalAdditionalCam
         }
 
         static void InitializeRenderingData(UniversalRenderPipelineAsset settings, ref CameraData cameraData, ref CullingResults cullResults,
-            bool anyPostProcessingEnabled, out RenderingData renderingData)
+            bool anyPostProcessingEnabled, out RenderingData renderingData, CommandBuffer cmd)
```
Comment (manuele-bonanno) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs:995]: nitpick: rearrange the parameters order so that the out parameter (renderingData) is last?

**Review by manuele-bonanno — CHANGES_REQUESTED** [review]
nice cleanup! Left some minor notes about small changes, overall looks good

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs (hunk 2 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs:166]: Yeah, i initially did this by passing the whole RenderingData, but then i thought that it's probably not necessary just for a single member of a struct. I thought that if there was a need for more info from renderingData - we could swap this then. I'm happy to change this to RenderingData if you think it's better that way

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs (hunk 1 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:92]: will investigate this

File: com.unity.render-pipelines.universal/Runtime/Debug/DebugHandler.cs (hunk 3 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/Debug/DebugHandler.cs:284]: will investigate this

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs (hunk 2 above)
Comment (manuele-bonanno) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs:166]: I did the same on my original prototype and I had this suggestion from Erik, so both are fine, but in general passing RenderingData is considered less restrictive

Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs:166]: sure. will change these calls to pass RendererData then

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs (hunk 10 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: fixed in 3939959

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 9 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs]: fixed in 3939959

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs (hunk 8 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs]: fixed in 3939959

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs (hunk 6 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs]: fixed in 3939959

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs (hunk 5 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs]: fixed in 3939959

File: com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs (hunk 2 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/2D/Renderer2D.cs]: fixed in 3939959

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs (hunk 1 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:92]: let's leave this to not affect the cmdbuffer syncing. these are hard to test issues as they could work on one platform but not the other and are hard to track and test

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs (hunk 4 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ForwardLights.cs:388]: The way I understand there's no way to remove the executes from the passes until we move to RendererList, otherwise our pass profiler markers will all be broken. Even after moving everything to command buffer, I think it's still safe to execute in between passes like is done [here](https://github.com/Unity-Technologies/Graphics/pull/6877/files#diff-b727bc7307e94c250d91774218394ee2a30e47117dce2fa62f42d8948af31a63R462-R463) as we don't know if a user is injecting a custom pass and running his own command buffer as they have access to the ScriptableRenderContext / CommandBufferPool.

File: com.unity.render-pipelines.universal/Runtime/Passes/DepthNormalOnlyPass.cs [hunk 11]
```diff
@@ -69,14 +69,8 @@ public override void OnCameraSetup(CommandBuffer cmd, ref RenderingData renderin
...
         public override void Execute(ScriptableRenderContext context, ref RenderingData renderingData)
         {
-            // NOTE: Do NOT mix ProfilingScope with named CommandBuffers i.e. CommandBufferPool.Get("name").
-            // Currently there's an issue which results in mismatched markers.
-            CommandBuffer cmd = CommandBufferPool.Get();
-            using (new ProfilingScope(cmd, ProfilingSampler.Get(URPProfileId.DepthNormalPrepass)))
+            using (new ProfilingScope(renderingData.commandBuffer, ProfilingSampler.Get(URPProfileId.DepthNormalPrepass)))
             {
-                context.ExecuteCommandBuffer(cmd);
-                cmd.Clear();
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/DepthNormalOnlyPass.cs:78]: Same here, you can't remove because there's a draw renderers being executed in the context that break the command buffer execution

File: com.unity.render-pipelines.universal/Runtime/Passes/DepthOnlyPass.cs [hunk 12]
```diff
@@ -77,22 +77,14 @@ public override void OnCameraSetup(CommandBuffer cmd, ref RenderingData renderin
...
         public override void Execute(ScriptableRenderContext context, ref RenderingData renderingData)
         {
-            // NOTE: Do NOT mix ProfilingScope with named CommandBuffers i.e. CommandBufferPool.Get("name").
-            // Currently there's an issue which results in mismatched markers.
-            CommandBuffer cmd = CommandBufferPool.Get();
-            using (new ProfilingScope(cmd, ProfilingSampler.Get(URPProfileId.DepthPrepass)))
+            using (new ProfilingScope(renderingData.commandBuffer, ProfilingSampler.Get(URPProfileId.DepthPrepass)))
             {
-                context.ExecuteCommandBuffer(cmd);
-                cmd.Clear();
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/DepthOnlyPass.cs:86]: And here

File: com.unity.render-pipelines.universal/Runtime/Passes/DrawObjectsPass.cs [hunk 13]
```diff
@@ -125,9 +123,6 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
                     : new Vector4(flipSign, 0.0f, 1.0f, 1.0f);
                 cmd.SetGlobalVector(ShaderPropertyId.scaleBiasRt, scaleBias);
 
-                context.ExecuteCommandBuffer(cmd);
-                cmd.Clear();
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/DrawObjectsPass.cs:129]: Same here, I believe this is needed because the [RenderingUtils.RenderObjectsWithError ](https://github.com/Unity-Technologies/Graphics/pull/6877/files#diff-e992d4c5e40f17f7ef44b9f3b4522526758d52ea042a50b5de6bc36aa42deac8L162)calls draw renderers in the context.

File: com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs [hunk 14]
```diff
@@ -112,20 +112,17 @@ public override void Configure(CommandBuffer cmd, RenderTextureDescriptor camera
 
         public override void Execute(ScriptableRenderContext context, ref RenderingData renderingData)
         {
-            CommandBuffer gbufferCommands = CommandBufferPool.Get();
-            using (new ProfilingScope(gbufferCommands, m_ProfilingSampler))
+            var cmd = renderingData.commandBuffer;
+            using (new ProfilingScope(cmd, m_ProfilingSampler))
             {
-                context.ExecuteCommandBuffer(gbufferCommands);
-                gbufferCommands.Clear();
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs:119]: Same here.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs [hunk 15]
```diff
@@ -292,16 +292,13 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
...
             if (m_IsFinalPass)
             {
-                var cmd = CommandBufferPool.Get();
                 using (new ProfilingScope(cmd, m_ProfilingRenderFinalPostProcessing))
                 {
                     RenderFinalPass(cmd, ref renderingData);
                 }
-
-                context.ExecuteCommandBuffer(cmd);
-                CommandBufferPool.Release(cmd);
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:304]: Might be worth checking if there isn't anything in the context breaking command buffer execution in post. I guess not as at it should do is Blit / DrawProcedural, but I'm not 100% sure.

File: com.unity.render-pipelines.universal/Runtime/Passes/DrawObjectsPass.cs (hunk 13 above)
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/DrawObjectsPass.cs:129]: This cannot be removed as well due to the RenderingUtils.RenderObjectsWithError.

File: com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs (hunk 14 above)
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs:119]: This might not be removed as well due to same reason.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs [hunk 16]
```diff
@@ -47,18 +47,19 @@ internal enum ImageUpscalingFilter
 
     public struct RenderingData
     {
+        public CommandBuffer commandBuffer;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:50]: Needs public API docs.

**Review by phi-lira — CHANGES_REQUESTED** [review]
Please also add changelog.

**Review by kennytann — APPROVED** [review]
Looks alright for the 2D Renderer

**Review by manuele-bonanno — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs (hunk 1 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:91]: I would think we don't need the execute & clear here or elsewhere most of the time. 

If I remember correctly, execute just takes a copy of the buf in the context and clear resets the local one.
Instead just collect everything into the buf and submit once. 

But it's completely ok to do this in another PR. 
There might have been some ownership issues with the cmdbufs.
Anyway, we just should add task for removing and testing these properly.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs [hunk 17]
```diff
@@ -47,18 +47,24 @@ internal enum ImageUpscalingFilter
 
     public struct RenderingData
     {
+        /// <summary>
+        /// Global ComandBuffer for the pipeline to be used instead of local command buffers inside ScriptableRenderPass.
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:51]:
```suggestion
        /// Global CommandBuffer for the pipeline to be used instead of local command buffers inside ScriptableRenderPass.
```

**Review by phi-lira — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs [hunk 18]
```diff
@@ -386,7 +386,7 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
                 this.DisableAllKeywords(cmd);
                 this.ReleaseRenderTextures(cmd);
                 context.ExecuteCommandBuffer(cmd);
-                CommandBufferPool.Release(cmd);
+                cmd.Clear();
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:389]: Is this really needed?

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs [hunk 19]
```diff
@@ -420,6 +420,7 @@ public override void Execute(ScriptableRenderContext context, ref RenderingData
 
                 this.DisableAllKeywords(cmd);
                 context.ExecuteCommandBuffer(cmd);
+                cmd.Clear();
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:423]: And here. I think we should try to get rid of as many as we can. 
Again, it's ok to do in another PR with proper testing and time.
If we need to call it for some reason, we should try to come up with some basic rules when it's necessary to guide other devs and users.

File: com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs (hunk 14 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs:119]: I don't even understand the old code.
1. We get a new buffer.
2. Add a profiling scope begin.
3. Copy the buffer (A) to context. (ctx: [A)
4. Clear the current buffer
---
5. Do stuff. Potentially a new cmd buf (B)
6. Copy the buffer(B) to context (ctx: [AB)
7. Clear local
---
8. Add profiling scope end to buffer C
9. ???? Execute/Submit somewhere possibly (ctx: [ABC)
10. ???? Might break? What if someone does a new buffer and executes that before this one.

Is this manually trying fix/manage profiling scopes?

Isn't this what we're trying to fix?
Just put everything in single buffer. The scope should work as it should automatically.

File: com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs [hunk 20]
```diff
@@ -112,20 +112,20 @@ public override void Configure(CommandBuffer cmd, RenderTextureDescriptor camera
...
                 // User can stack several scriptable renderers during rendering but deferred renderer should only lit pixels added by this gbuffer pass.
                 // If we detect we are in such case (camera is in overlay mode), we clear the highest bits of stencil we have control of and use them to
                 // mark what pixel to shade during deferred pass. Gbuffer will always mark pixels using their material types.
                 if (m_DeferredLights.IsOverlay)
                 {
-                    m_DeferredLights.ClearStencilPartial(gbufferCommands);
-                    context.ExecuteCommandBuffer(gbufferCommands);
-                    gbufferCommands.Clear();
+                    m_DeferredLights.ClearStencilPartial(cmd);
+                    context.ExecuteCommandBuffer(cmd);
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/Passes/GBufferPass.cs:127]: I don't think these are needed as long as the ClearStencilPartial uses the same buf.

**Review by eh-unity — APPROVED** [review]
Looks like a good start. I think we should try to remove all the Execute&Clears if possible. That can be done in another PR(s). We should take time to test everything.

Maybe there was some issues with cmdbuf ownership. I would imagine everything could be just 1 buf without exec&clear as long as the buf doesn't have a single state for something and doesn't modify the context etc.

If there is this sort of "non-versioned state", we should define some basic rules how to use the cmdbuf currently and document that for other devs and users.

File: com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs (hunk 1 above)
Comment (jonuuukas) [reply com.unity.render-pipelines.universal/Runtime/2D/Passes/Render2DLightingPass.cs:91]: there are several issues with removing these calls, the main culprit being ProfilingScopes and clashes with Context calls. I guess we could do a lot of these calls cleanup when we change context.DrawRenderers and similar calls with cmd.DrawRendererLists. I suggest now to play it safe and keep these calls, and cleanup in a single sweep when we can

**Review by simon-engelbrecht-soerensen — APPROVED** [review]
Checked URP Rendering Examples and Boat Attack up against master and found no discrepancies.
